File: Python/0394-decode-string.py
# Tags: String, Stack
# Time:  O(n)
# Space: O(n)
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        cur_num = 0
        cur_string = ''
        for c in s:
            if c == '[':
                # record cur string
                stack.append(cur_string)
                # record cur num
                stack.append(cur_num)
                # reset
                cur_string = ''
                cur_num = 0

            elif c == ']':
                num = stack.pop()
                prev_string = stack.pop()
                cur_string = prev_string + num*cur_string
            elif c.isdigit():
                cur_num = cur_num*10 + int(c)
                # or cur_num*10 + ord(c) - ord('0')
            else:
                # isalpha()
                cur_string += c

        return cur_string

File: C++/0394-decode-string.cpp
// Tags: String Stack Amazon
// Time:  O(n)
// Space: O(n)
// Similar to 0224-basic-calculator
class Solution {
public:
    string decodeString(string s) {
        stack<string> st;
        string cur;
        int num = 0;
        for (int i = 0; i < s.length(); ++i) {
            if (isdigit(s[i])) {
                num = 10*num + (s[i] - '0');
            } else if (s[i] == '[') {
                st.push(cur);
                st.push(to_string(num));
                cur = "";
                num = 0;
            } else if (s[i] == ']') {
                int times = stoi(st.top()); st.pop();
                string prev = st.top(); st.pop();
                while (times-- > 0) {
                    prev += cur;
                }
                cur = prev;
            } else {
                cur += s[i];
            }
        }
        return cur;
    }
};
